### DISCUSSION TITLE
Helm Repo To Package Repo Conversion Tool + Extra Tooling

### DISCUSSION THREAD

**Discussion by vrabbi** [vmware-tanzu/community-edition#2466] [Show and tell]
TLDR: helpful links:
https://github.com/TeraSky-Ltd/helm-to-carvel-conversion-tool
https://github.com/vrabbi/tanzu-helper-scripts
https://github.com/vrabbi/tkgm-customizations/tree/master/carvel-packages/kube-vip-package

I have been working on a tool for automating the conversion of helm repositories into carvel package repositories and also utilizing the carvel toolset to allow for airgapped scenarios.

originally this was built for bitnami helm charts but has been expanded since to support any helm repository following community standards of having images not hard coded in the templates rather parameterized in the values.yaml file.

the Git Repo for the new and updated tool is https://github.com/TeraSky-Ltd/helm-to-carvel-conversion-tool

This allows for 2 modes:
1. Non Air Gapped - Takes very little storage in your OCI registry and is fuly automated. only helm chart files and package metadata files are saved in the package bundle and container images will be deployed from the default registries.
2. Air Gapped - Requires running a few commands and importing the tar file to your air gapped environment. this will take considerable space as all container images will also be copied into your air gapped environment. for example the bitnami helm repo when saving the last 4 versions of all charts takes 57 GB of space in the destination repo.

this has been tested against the following public helm repositories:
1. bitnami Repo - https://charts.bitnami.com/bitnami
2. Tanzu Application Catalog Trial Repo - https://charts.trials.tac.bitnami.com/demo
3. Hashicorp Repo - https://helm.releases.hashicorp.com
4. K8s at home - https://k8s-at-home.com/charts/

The only requirements for the helm repo in order for this to work are:
1. The images must not be hardcoded in order to get airgapped support - the packaging will work but no images will be packaged along with the helm chart that are not in the values.yaml
2. the images in the values.yaml must have at minimum a repository and tag field and also support a registry field. those values are utilized to build the image references. any other formatting is currently unsupported.
3. the helm charts must be helm v3 charts and not helm v2 - shouldnt be an issue as helm v2 is EOL.
4. the helm repo must allow unauthenticated access currently. - if support for auth is needed let me know and it can be added but this is currently low priority

Any comments or feedback are appreciated!

I have also created a Kube-Vip Package for ease of installation which can be found at the following link - https://github.com/vrabbi/tkgm-customizations/tree/master/carvel-packages/kube-vip-package

I have also created a set of helper scripts for Tanzu (TCE and TKGm):
1. A script to help generate a values file for the tanzu package install command based on the values schema with all defaults. this can be used with the packages that come with TCE or the ones generated using this tool or really any carvel package as long as there is a values schema in the package.
2. A script to export all tanzu kubeconfigs in 3 modes (admin,pinniped and browserless pinniped). the kubeconfigs are exported as seperate kubeconfigs per cluster as well as in a combined kubeconfig file for all clusters to meet whatever use case is needed.

The scripts can be found on my github at the following link: https://github.com/vrabbi/tanzu-helper-scripts
